packages/mui-base/src/SnackbarUnstyled/useSnackbar.ts

```diff
@@ -0,0 +1,183 @@
+import * as React from 'react';
+import { unstable_useEventCallback as useEventCallback } from '@mui/utils';
+import {
+  UseSnackbarParameters,
+  SnackbarCloseReason,
+  UseSnackbarRootSlotProps,
+} from './useSnackbar.types';
+import extractEventHandlers from '../utils/extractEventHandlers';
+
+/**
+ * The basic building block for creating custom snackbar.
+ *
+ * Demos:
+ *
+ * - [Snackbar](https://mui.com/base/react-snackbar/)
+ */
+export default function useSnackbar(parameters: UseSnackbarParameters) {
+  const {
+    autoHideDuration = null,
+    disableWindowBlurListener = false,
+    onClose,
+    open,
+    ref,
+    resumeHideDuration,
+  } = parameters;
+
+  const timerAutoHide = React.useRef<ReturnType<typeof setTimeout>>();
+
+  React.useEffect(() => {
+    if (!open) {
+      return undefined;
+    }
+
+    /**
+     * @param {KeyboardEvent} nativeEvent
+     */
+    function handleKeyDown(nativeEvent: KeyboardEvent) {
+      if (!nativeEvent.defaultPrevented) {
+        // IE11, Edge (prior to using Blink?) use 'Esc'
+        if (nativeEvent.key === 'Escape' || nativeEvent.key === 'Esc') {
+          // not calling `preventDefault` since we don't know if people may ignore this event e.g. a permanently open snackbar
+          if (onClose) {
+            onClose(nativeEvent, 'escapeKeyDown');
+          }
+        }
+      }
+    }
+
+    document.addEventListener('keydown', handleKeyDown);
+
+    return () => {
+      document.removeEventListener('keydown', handleKeyDown);
+    };
+  }, [open, onClose]);
+
+  const handleClose = useEventCallback(
+    (event: Event | React.SyntheticEvent<any, Event> | null, reason: SnackbarCloseReason) => {
+      if (onClose) {
+        onClose(event, reason);
+      }
```

Comment:
We tend to use the optional chaining operator across the new codebase.

```suggestion
      onClose?.(event, reason);    
```

Please update the other cases below as well.